### PR TITLE
Node Restriction

### PR DESCRIPTION
**Pull request by xpivarc**
### What this PR does
Introduce NodeRestriction feature gate. This enables
us to check if a virt-handler request is authorized to
modify VMI.

This feature requires following Kubernetes feature gate
ServiceAccountTokenPodNodeInfo. The feature gate is available
in 1.30 as Beta.

Fixes #

### Release note

```release-note
By enabling NodeRestriction feature gate, Kubevirt now authorize virt-handler's requests to modify VMs.
```

### CODE REVIEW

**Comment by Barakmor1** [comment]
@xpivarc I'm not sure i understand the value of the FeatureGate can you please elaborate?

**Comment by RamLavi** [comment]
@xpivarc I think we should wait until this k8s feature gate is GA.

**Comment by xpivarc** [comment]
> @xpivarc I think we should wait until this k8s feature gate is GA.

A feature gate is provided to opt-in when possible. I don't see a reason to hold a security feature that can be opted-out again in case it will not make it out of Beta(highly unlikely).

**Comment by xpivarc** [comment]
> Also, wouldn't it make more sense to protect VMI modification using Validating Admission Policy?

Imho, there is no benefit. We already have a webhook and native code will always be faster than interpreting the CEL.

**Comment by fossedihelm** [comment]
Keep in mind possible unwanted behavior, running the new e2e tests in parallel, if https://github.com/kubevirt/kubevirt/pull/12033 is merged first

**Comment by xpivarc** [comment]
> Keep in mind possible unwanted behavior, running the new e2e tests in parallel, if #12033 is merged first

I rely on the test to fail.

File: automation/test.sh
Comment (fossedihelm) [inline automation/test.sh:459]:
```suggestion
  if [[ ! $TARGET =~ k8s-1\.3[0-9].* ]]; then
```
It should not happen but the previous regex matches k8s-1.3.0 :)

Comment (xpivarc) [reply automation/test.sh:459]: Since this is temporary and the target is managed by us, I don't think we need to be defensive, in worst case it will fail on the test.

Comment (fossedihelm) [reply automation/test.sh:459]: Sure, thanks!

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter_test.go
Comment (jean-edouard) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter_test.go:141]: This far-indented and hard-to-read block is repeated 3 times (unless I missed diffs), might be worth creating a function for it.

**Review by jean-edouard — COMMENTED** [review]
Looks good, few suggestions in-line. Thanks

File: automation/test.sh
Comment (xpivarc) [reply automation/test.sh:459]: I doubt we are going to change the naming in the future. Most of the time, @dhiller, @brianmcarey  or me are creating the new lanes (and automation). So I can change it but I am pretty confident here.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go
Comment (xpivarc) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go:71]: I'll add a comment

Comment (jean-edouard) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go:76]: Don't we want to catch targetNode changes from x to y as well?

**Review by jean-edouard — COMMENTED** [review]
/approve

**Review by fossedihelm — APPROVED** [review]

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go
Comment (xpivarc) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go:76]: The first check makes sure that the target is only source or target. This is valid but we can improve later

File: tests/infrastructure/security.go
Comment (jean-edouard) [inline tests/infrastructure/security.go:103]:
```suggestion
			ContainSubstring("Node restriction, virt-handler is only allowed to modify VMIs it owns"),
```

**Comment by jean-edouard** [comment]
/lgtm
Re-LGTM-ing after typo fix

File: pkg/virt-api/webhooks/utils.go
Comment (orelmisan) [reply pkg/virt-api/webhooks/utils.go:104]: @xpivarc @jean-edouard please see https://github.com/kubevirt/kubevirt/pull/11931.

**Comment by EdDev** [comment]
Was this new FG discussed & negotiated as part of an enhancement proposal or an alternative discussion?

**Comment by xpivarc** [comment]
> Was this new FG discussed & negotiated as part of an enhancement proposal or an alternative discussion?

It was discussed within approvers.

**Comment by fabiand** [comment]
Thats fine, xpivarc.

But for reference please just record any discussion on PRs and alike.

**Comment by EdDev** [comment]
AFAIK this option has been mentioned in the node restriction proposal [1] [2] , so I guess this is a follow up action.

Can you please share here the reason why we need a FG instead of just detecting the availability of this feature from K8S? I.e. if it exists, this addition can be added and if it is not available, not to introduce it.
I guess in the same manner the existing non FGed option was introduced.

[1] https://github.com/kubevirt/community/pull/246/files#diff-844e2d53097784148c80d78a29bb4f16f925585d886d00d49f2ad83f8bb6b486R88
[2] https://github.com/kubevirt/community/pull/246/files#diff-844e2d53097784148c80d78a29bb4f16f925585d886d00d49f2ad83f8bb6b486R345

**Comment by xpivarc** [comment]
> AFAIK this option has been mentioned in the node restriction proposal [1] [2] , so I guess this is a follow up action.

 Not sure what "option" is referring to here. I will guess you are referring to the k8s feature gate. The feature gate is needed for the https://github.com/kubevirt/community/pull/246 only partially as you pointed out. 
 
This PR does have some relationship with the proposal. It does refer or actually use the k8s feature gate which is referenced in the proposal. It fixes a trampoline pod problem as well as the proposal. In what they differ is the object that's usage is being restricted. VMI vs Node.


> Can you please share here the reason why we need a FG instead of just detecting the availability of this feature from K8S? I.e. if it exists, this addition can be added and if it is not available, not to introduce it. I guess in the same manner the existing non FGed option was introduced.

As far as I know there is no easy way to detect a specific feature gate is available in the cluster (+ what is the maturity). The difference here with the proposal is that in the proposal we rely on an API that is actually versioned and easily probed.

> 
> [1] https://github.com/kubevirt/community/pull/246/files#diff-844e2d53097784148c80d78a29bb4f16f925585d886d00d49f2ad83f8bb6b486R88 [2] https://github.com/kubevirt/community/pull/246/files#diff-844e2d53097784148c80d78a29bb4f16f925585d886d00d49f2ad83f8bb6b486R345

**Comment by EdDev** [comment]
> As far as I know there is no easy way to detect a specific feature gate is available in the cluster (+ what is the maturity). The difference here with the proposal is that in the proposal we rely on an API that is actually versioned and easily probed.

Well, an option/feature can be detected by the CR version, by the ability to apply a configuration without errors and even by assuming the spec is ignored if some parts of it are unsupported.
Have these options been considered?

To clarify, unlike a regular feature that is exposed to our users as evaluation/experiment, this is an internal backend implementation that comes to resolve a security issues. There is no promise given to users from an API or behavioral aspect, we just say something like "we'll try our best to make it as secured as possible". If we can do that without placing the burden on the user, I think it is preferable.
I do not know if this is possible in this case or not, but it would have been great if it was considered.

**Comment by xpivarc** [comment]
> Well, an option/feature can be detected by the CR version, by the ability to apply a configuration without errors and even by assuming the spec is ignored if some parts of it are unsupported. Have these options been considered?
> 

Can you elaborate?

**Comment by xpivarc** [comment]
> > > Well, an option/feature can be detected by the CR version, by the ability to apply a configuration without errors and even by assuming the spec is ignored if some parts of it are unsupported. Have these options been considered?
> > 
> > 
> > Can you elaborate?
> 
>     * Detect by error: Try to apply a policy spec that uses the capability that is protected by a K8S FG. If there is an error that is raised, our code can interpret that this feature not available in the cluster and avoid using it.

There is nothing to apply unfortunately and the Kubernetes feature is Beta. I am not convinced we should use Beta features by default without any consent of the users.

> 
>     * Ignored by the implementation: Similar to the error scenario before, but without an error. The implementation does not raise an error and the unsupported policy settings are ignored.

This could work but I am not big fan of just ignoring stuff, we should error out but as mentioned above it could be possible to do only some kind of crazy runtime check. Because this is security fix/feature I don't feel comfortable to ignore feature nor to do flaky runtime checks. For example there could be a SA not bind to node even if it should be... From security POV I think it is much advisable to reject request that do not comply...
> 
> 
> I would expect to see an error, but both options are manageable by the code that uses it IMO.
> 
> I also do not see a problem to query the cluster for the FG setting and act upon that information.

I am more than happy if you show me how can we reliably query this.

**Comment by RamLavi** [comment]
> > > > Well, an option/feature can be detected by the CR version, by the ability to apply a configuration without errors and even by assuming the spec is ignored if some parts of it are unsupported. Have these options been considered?
> > > 
> > > 
> > > Can you elaborate?
> > 
> > 
> > ```
> > * Detect by error: Try to apply a policy spec that uses the capability that is protected by a K8S FG. If there is an error that is raised, our code can interpret that this feature not available in the cluster and avoid using it.
> > ```
> 
> There is nothing to apply unfortunately and the Kubernetes feature is Beta. I am not convinced we should use Beta features by default without any consent of the users.
> 
> > ```
> > * Ignored by the implementation: Similar to the error scenario before, but without an error. The implementation does not raise an error and the unsupported policy settings are ignored.
> > ```
> 
> This could work but I am not big fan of just ignoring stuff, we should error out but as mentioned above it could be possible to do only some kind of crazy runtime check. Because this is security fix/feature I don't feel comfortable to ignore feature nor to do flaky runtime checks. For example there could be a SA not bind to node even if it should be... From security POV I think it is much advisable to reject request that do not comply...
> 
> > I would expect to see an error, but both options are manageable by the code that uses it IMO.
> > I also do not see a problem to query the cluster for the FG setting and act upon that information.
> 
> I am more than happy if you show me how can we reliably query this.

The only way I can think of is to check the `kube-apiserver-<node-name>` pod in `Spec.Containers[0].Args/Command` for the feature-gates. I understand that on OCP it works differently though.

In any case I agree that ignoring errors is not a good approach, but what about a "prohibit if the `authentication.kubernetes.io/node-name` extra field **exists**, and do nothing if it is not there"? When the feature goes GA we can make it more strict, but it won't require any proactive work from the cluster-admin. 

***
In general though, I think the question here is whether it is worth introducing a beta k8s functionality _at all_ - for the sake of a "security" threat. As in all things - it depends on the case, as security threats come in a variety of impact severity.

IMO - the security impact in this case does not justify inserting a beta k8s feature. 
Furthermore, adding our own FG to wrap the already existing K8s FG will only render it to be "never used". It is not the cluster-admin's job to opt-in/register to security fixes, and we should not delegate this responsibility to them.

So the question I want to raise is - can't we wait until this functionality is GA? 
I know we differ in opinions here so I suggest we ask more people and get more thoughts. @fabiand WDYT?

**Comment by xpivarc** [comment]
> > > > > Well, an option/feature can be detected by the CR version, by the ability to apply a configuration without errors and even by assuming the spec is ignored if some parts of it are unsupported. Have these options been considered?
> > > > 
> > > > 
> > > > Can you elaborate?
> > > 
> > > 
> > > ```
> > > * Detect by error: Try to apply a policy spec that uses the capability that is protected by a K8S FG. If there is an error that is raised, our code can interpret that this feature not available in the cluster and avoid using it.
> > > ```
> > 
> > 
> > There is nothing to apply unfortunately and the Kubernetes feature is Beta. I am not convinced we should use Beta features by default without any consent of the users.
> > > ```
> > > * Ignored by the implementation: Similar to the error scenario before, but without an error. The implementation does not raise an error and the unsupported policy settings are ignored.
> > > ```
> > 
> > 
> > This could work but I am not big fan of just ignoring stuff, we should error out but as mentioned above it could be possible to do only some kind of crazy runtime check. Because this is security fix/feature I don't feel comfortable to ignore feature nor to do flaky runtime checks. For example there could be a SA not bind to node even if it should be... From security POV I think it is much advisable to reject request that do not comply...
> > > I would expect to see an error, but both options are manageable by the code that uses it IMO.
> > > I also do not see a problem to query the cluster for the FG setting and act upon that information.
> > 
> > 
> > I am more than happy if you show me how can we reliably query this.
> 
> The only way I can think of is to check the `kube-apiserver-<node-name>` pod in `Spec.Containers[0].Args/Command` for the feature-gates. I understand that on OCP it works differently though.

This is impractical and we should not have permission to the namespace of Kubernetes system.

> 
> In any case I agree that ignoring errors is not a good approach, but what about a "prohibit if the `authentication.kubernetes.io/node-name` extra field **exists**, and do nothing if it is not there"? When the feature goes GA we can make it more strict, but it won't require any proactive work from the cluster-admin.

I am not sure but to me this is equivalent of ignoring errors.

> In general though, I think the question here is whether it is worth introducing a beta k8s functionality _at all_ - for the sake of a "security" threat. As in all things - it depends on the case, as security threats come in a variety of impact severity.
> 
> IMO - the security impact in this case does not justify inserting a beta k8s feature. Furthermore, adding our own FG to wrap the already existing K8s FG will only render it to be "never used". It is not the cluster-admin's job to opt-in/register to security fixes, and we should not delegate this responsibility to them.

Here I agree that opting in for security fixes is really bad practice and we should not make it a custom in our project. Now the reason why I saw a need for opt-in here is that there is not good "detection" for this to do it auto-magically. I do think that providing a security fix is worth the effort even if the fix is gated. 

> 
> So the question I want to raise is - can't we wait until this functionality is GA? Is the severity of the security threat worth relying on something that is not yet approved as stable by it's creators? I know we differ in opinions here so I suggest we ask more people and get more thoughts. @fabiand WDYT?

**Pull request merged**
